src/main/java/com/gregtechceu/gtceu/integration/map/ftbchunks/veins/ore/OreVeinIcon.java

```diff
@@ -0,0 +1,152 @@
+import net.minecraft.world.inventory.InventoryMenu;
+import net.minecraft.world.phys.Vec3;
+
+import com.mojang.blaze3d.platform.InputConstants;
+import com.mojang.blaze3d.systems.RenderSystem;
+import dev.ftb.mods.ftbchunks.api.client.icon.MapIcon;
+import dev.ftb.mods.ftbchunks.api.client.icon.MapType;
+import dev.ftb.mods.ftbchunks.client.map.MapDimension;
+import dev.ftb.mods.ftbchunks.client.map.WaypointImpl;
+import dev.ftb.mods.ftbchunks.client.map.WaypointType;
+import dev.ftb.mods.ftblibrary.ui.BaseScreen;
+import dev.ftb.mods.ftblibrary.ui.input.Key;
+import dev.ftb.mods.ftblibrary.ui.input.MouseButton;
+import dev.ftb.mods.ftblibrary.util.TooltipList;
+
+import java.util.function.Function;
+
+public class OreVeinIcon implements MapIcon {
+
+    protected final GeneratedVeinMetadata veinMetadata;
+    protected String name = null;
+
+    public OreVeinIcon(GeneratedVeinMetadata veinMetadata) {
+        this.veinMetadata = veinMetadata;
+    }
+
+    public boolean isEnabled() {
+        return FTBChunksPlugin.getInstance().getOptions().showLayer("ore_veins") &&
+                !(veinMetadata.depleted() && FTBChunksPlugin.getInstance().getOptions().hideDepleted());
+    }
+
+    public String getName() {
+        return name == null ? name = OreRenderLayer.getName(veinMetadata).getString() : name;
+    }
+
+    public Material getMaterial() {
+        Material firstMaterial = null;
+        if (!veinMetadata.definition().indicatorGenerators().isEmpty()) {
+            var blockOrMaterial = veinMetadata.definition().indicatorGenerators().get(0).block();
+            firstMaterial = blockOrMaterial == null ? null : blockOrMaterial.map(
+                    state -> {
+                        var matStack = ChemicalHelper.getMaterial(state.getBlock());
+                        return matStack == null ? null : matStack.material();
+                    },
+                    Function.identity());
+        }
+        if (firstMaterial == null) {
+            firstMaterial = veinMetadata.definition().veinGenerator().getAllMaterials().get(0);
+        }
+        return firstMaterial;
+    }
+
+    @Override
+    public Vec3 getPos(float v) {
+        return veinMetadata.center().getCenter();
+    }
+
+    @Override
+    public boolean onMousePressed(BaseScreen baseScreen, MouseButton mouseButton) {
+        if (!isEnabled()) {
+            return false;
+        }
+        MapDimension.getCurrent()
+                .ifPresent(mapDimension -> FTBChunksPlugin.getInstance().getClientAPI()
+                        .getWaypointManager(mapDimension.dimension)
+                        .ifPresent(waypointManager -> {
+                            var waypoint = new WaypointImpl(WaypointType.DEFAULT, mapDimension, veinMetadata.center());
+                            if (!waypointManager.getAllWaypoints().contains(waypoint)) {
+                                waypointManager.addWaypointAt(veinMetadata.center(), getName())
```

Comment:
The waypointManager doesn't automatically refresh the current screen, so clicking the ore icons looks like it doesn't do anything. Adding a call to `baseScreen.openGui();` makes the screen refresh after clicking.